### ISSUE TITLE
Pandas TA strategy method goes into infinite loop

### ISSUE THREAD

**Issue by twopirllc** [twopirllc/pandas-ta#176]
@twopirllc I was trying to run the below code on a dataframe

```python
df.ta.strategy("Momentum")
print(df)
```
It goes into infinite loop and consumes all CPU on my windows 10 laptop.
Then I need to kill all VSCodium and python exes.
Could you please help?

_Originally posted by @rahulmr in https://github.com/twopirllc/pandas-ta/issues/138#issuecomment-751864255_

**Comment by twopirllc** [comment]
Hello @rahulmr,

You are leaving a lot out of context for me to help you at this moment. If you could provide me more code to work with and some sample data that you are using, then we can move further.

Thanks,
KJ

**Comment by rahulmr** [comment]
code is very simple as below, I am not sure whether I am doing it in right way. But why it goes to infinite loop is what I am not able to understand. If there is some problem it must throw an error.

```
import yfinance as yf
import pandas_ta as ta


sbin = yf.Ticker("SBIN.NS")
df = sbin.history(period="3d", interval="15m")
df.ta.strategy("Momentum")
```

The output is as below, somehow tried to copy it.

```
File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 114, in _main
    prepare(preparation_data)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 225, in prepare
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 105, in spawn_main
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 277, in _fixup_main_from_path
    exitcode = _main(fd)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 114, in _main
    run_name="__mp_main__")
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 263, in run_path
    prepare(preparation_data)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 225, in prepare
    pkg_name=pkg_name, script_name=fname)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 96, in _run_module_code
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 277, in _fixup_main_from_path
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 85, in _run_code
    run_name="__mp_main__")
    exec(code, run_globals)
  File "D:\trading\0sas_algo\crash.py", line 7, in <module>
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 263, in run_path
    df.ta.strategy("Momentum")
    pkg_name=pkg_name, script_name=fname)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 96, in _run_module_code
  File "D:\trading\0sas_algo\senv\lib\site-packages\pandas_ta\core.py", line 633, in strategy
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 85, in _run_code
    pool = Pool(self.cores)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 119, in Pool
    exec(code, run_globals)
  File "D:\trading\0sas_algo\crash.py", line 7, in <module>
    context=self.get_context())
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 176, in __init__
    df.ta.strategy("Momentum")
  File "D:\trading\0sas_algo\senv\lib\site-packages\pandas_ta\core.py", line 633, in strategy
    self._repopulate_pool()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 241, in _repopulate_pool
    pool = Pool(self.cores)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 119, in Pool
    w.start()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\process.py", line 112, in start
    context=self.get_context())
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 176, in __init__
    self._popen = self._Popen(self)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 322, in _Popen
    self._repopulate_pool()
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 241, in _repopulate_pool
    w.start()
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\process.py", line 112, in start
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    self._popen = self._Popen(self)
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 322, in _Popen
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 114, in _main
    prepare(preparation_data)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 225, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 277, in _fixup_main_from_path
    run_name="__mp_main__")
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\trading\0sas_algo\crash.py", line 7, in <module>
    df.ta.strategy("Momentum")
  File "D:\trading\0sas_algo\senv\lib\site-packages\pandas_ta\core.py", line 633, in strategy
    pool = Pool(self.cores)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 119, in Pool
    context=self.get_context())
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 176, in __init__
    self._repopulate_pool()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 241, in _repopulate_pool
    w.start()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\process.py", line 112, in start
    self._popen = self._Popen(self)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 322, in _Popen
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 114, in _main
    prepare(preparation_data)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 225, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 277, in _fixup_main_from_path
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    run_name="__mp_main__")
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 105, in spawn_main
  File "D:\trading\0sas_algo\crash.py", line 7, in <module>
    exitcode = _main(fd)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 114, in _main
    df.ta.strategy("Momentum")
  File "D:\trading\0sas_algo\senv\lib\site-packages\pandas_ta\core.py", line 633, in strategy
    prepare(preparation_data)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 225, in prepare
    pool = Pool(self.cores)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 119, in Pool
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 277, in _fixup_main_from_path
    context=self.get_context())
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 176, in __init__
    run_name="__mp_main__")
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 263, in run_path
    self._repopulate_pool()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 241, in _repopulate_pool
    pkg_name=pkg_name, script_name=fname)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 96, in _run_module_code
    w.start()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\process.py", line 112, in start
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\runpy.py", line 85, in _run_code
    self._popen = self._Popen(self)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 322, in _Popen
    exec(code, run_globals)
  File "D:\trading\0sas_algo\crash.py", line 7, in <module>
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    df.ta.strategy("Momentum")
  File "D:\trading\0sas_algo\senv\lib\site-packages\pandas_ta\core.py", line 633, in strategy
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    pool = Pool(self.cores)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 119, in Pool
    _check_not_importing_main()
    context=self.get_context())
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 176, in __init__
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    self._repopulate_pool()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\pool.py", line 241, in _repopulate_pool
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    w.start()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\process.py", line 112, in start
    self._popen = self._Popen(self)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\context.py", line 322, in _Popen
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
```

**Comment by rahulmr** [comment]
Mine is windows 10 and I am using VSCodium IDE, Python 3.7.9 64 bit

**Comment by rahulmr** [comment]
Again tried with python 3.9.1 and still getting same issue. Control + C is also not working.
Anyways this is not high priority for me. I will concentrate more on the other stuff. Will continue there.
Let me know if you are able to reproduce this issue at your end. I can try to uploaded a video if I get time.

**Comment by twopirllc** [comment]
Thanks for bringing the Issue here @rahulmr. It makes is easier for other users to find this similar Issue.

> Anyways this is not high priority for me. I will concentrate more on the other stuff. Will continue there. Let me know if you are able to reproduce this issue at your end. I can try to uploaded a video if I get time.

Also, what about the other code on the previous Issue? Or is that code not relevant?

Thanks,
KJ

**Comment by rahulmr** [comment]
Seems like issue with Windows 10, not sure though. I will try to execute the same code on other Windows VMs and confirm.

**Comment by rahulmr** [comment]
> Also, what about the other code on the previous Issue? Or is that code not relevant?

No it is not relevant here. That issue is actually not an issue but help I need on how to work with pandas_ta. We will discuss it there.

**Comment by twopirllc** [comment]
I haven't tested it on Windows 10 yet. But will soon.

**Comment by rahulmr** [comment]
Yes, same issue on another windows machines as well. May be you can try Vultr Cloud for windows. It seems to be related with the multiprocessing module. Spwan.py

**Comment by twopirllc** [comment]
@rahulmr,

Interesting. I am not having that issue with my Windows 10 laptop either. 

My Windows 10 results:

![pandas-ta-strategy-win10-ipython](https://user-images.githubusercontent.com/18198392/103250194-965a8900-4927-11eb-88fd-22bc5c828611.png)

Are you strictly using just those 4 lines of code or are you all trying to set the number of cores prior to calling the _df.ta.strategy_ method?

Thanks,
KJ

**Comment by rahulmr** [comment]
> @rahulmr,
> 
> Interesting. I am not having that issue with my Windows 10 laptop either.
> 
> My Windows 10 results:
> 
> ![pandas-ta-strategy-win10-ipython](https://user-images.githubusercontent.com/18198392/103250194-965a8900-4927-11eb-88fd-22bc5c828611.png)
> 
> Are you strictly using just those 5 lines of code or are you all trying to set the number of cores prior to calling the _df.ta.strategy_ method?
> 
> Thanks,
> KJ

That's surprising for me. I will try your method from python prompt. Could you please try by running a script using `python crash.py`

**Comment by rahulmr** [comment]
This is great - some kind of tricky issue for me.

It is working from python prompt.
```
>>> sbin = yf.Ticker("SBIN.NS")
>>> df = sbin.history(period="3d", interval="15m")
>>> df.ta.strategy("Momentum")
>>> print(df)
                                 open        high         low       close   volume  Dividends  Stock Splits  ...  FISHERTs_9_1  INERTIA_20_14      K_9_3      D_9_3      J_9_3  KST_10_15_20_30_10_10_10_15       KSTs_9
datetime                                                                                                     ...
2020-12-24 09:15:00+05:30  265.100006  266.549988  264.700012  266.299988  1653388          0             0  ...           NaN            NaN        NaN        NaN        NaN                          NaN          NaN
2020-12-24 09:30:00+05:30  266.299988  266.649994  265.500000  266.299988  1346556          0             0  ...           NaN            NaN        NaN        NaN        NaN                          NaN          NaN
2020-12-24 09:45:00+05:30  266.299988  266.450012  265.500000  266.100006   914542          0             0  ...           NaN            NaN        NaN        NaN        NaN                          NaN          NaN
2020-12-24 10:00:00+05:30  266.100006  267.649994  266.100006  267.399994  1594715          0             0  ...           NaN            NaN        NaN        NaN        NaN                          NaN          NaN
2020-12-24 10:15:00+05:30  267.299988  267.799988  266.649994  266.799988  1050138          0             0  ...           NaN            NaN        NaN        NaN        NaN                          NaN          NaN
...                               ...         ...         ...         ...      ...        ...           ...  ...           ...            ...        ...        ...        ...                          ...          ...
2020-12-29 14:15:00+05:30  278.000000  278.350006  277.049988  277.399994  1333187          0             0  ...      0.900915      36.517981  66.863205  64.981378  70.626858                  1306.427229  1751.919538
2020-12-29 14:30:00+05:30  277.500000  277.899994  276.500000  277.750000  1194604          0             0  ...      0.996842      38.154401  67.492137  65.818298  70.839814                  1252.954426  1640.715420
2020-12-29 14:45:00+05:30  277.649994  277.850006  276.850006  276.950012   841302          0             0  ...      0.887071      37.971401  57.302575  62.979723  45.948279                  1186.942713  1539.359095
2020-12-29 15:00:00+05:30  277.000000  278.399994  277.000000  277.950012  2298641          0             0  ...      0.700636      39.792620  59.254517  61.737988  54.287576                  1154.360291  1451.367812
2020-12-29 15:15:00+05:30  277.950012  278.450012  277.250000  277.750000  2537262          0             0  ...      0.442987      41.414788  56.169678  59.881885  48.745265                  1095.710145  1367.014473

[75 rows x 62 columns]

```

**Comment by rahulmr** [comment]
Let me try it on Linux container. And then I will add the updates.

**Comment by rahulmr** [comment]
Great it is working on my docker container
Details here:
```
root@node54310-pandas-r:~/sas# cat sbin.py
import yfinance as yf
import pandas_ta as ta


sbin = yf.Ticker("SBIN.NS")
df = sbin.history(period="3d", interval="15m")
df.ta.strategy("Momentum", verbose=True)
print(df)

root@node54310-pandas-r:~/sas# python sbin.py
[+] Strategy: Momentum
[i] Indicator arguments: {'append': True}
[i] Excluded[10]: above, above_value, below, below_value, cross, cross_value, long_run, short_run, trend_return, vp
[i] Multiprocessing: 3 of 3 cores.
[i] Total indicators: 34
[i] Columns added: 55
                                 open        high         low       close   volume  Dividends  Stock Splits  ...  SQZ_ON  SQZ_OFF  SQZ_NO  STOCHk_14_3_3  STOCHd_14_3_3  STOCHRSIk_14_14_3_3  STOCHRSId_14_14_3_3
datetime                                                                                                     ...                                                                                              
2020-12-24 09:15:00+05:30  265.100006  266.549988  264.700012  266.299988  1653388          0             0  ...       0        0       1            NaN            NaN                  NaN                  NaN
2020-12-24 09:30:00+05:30  266.299988  266.649994  265.500000  266.299988  1346556          0             0  ...       0        0       1            NaN            NaN                  NaN                  NaN
2020-12-24 09:45:00+05:30  266.299988  266.450012  265.500000  266.100006   914542          0             0  ...       0        0       1            NaN            NaN                  NaN                  NaN
2020-12-24 10:00:00+05:30  266.100006  267.649994  266.100006  267.399994  1594715          0             0  ...       0        0       1            NaN            NaN                  NaN                  NaN
2020-12-24 10:15:00+05:30  267.299988  267.799988  266.649994  266.799988  1050138          0             0  ...       0        0       1            NaN            NaN                  NaN                  NaN
...                               ...         ...         ...         ...      ...        ...           ...  ...     ...      ...     ...            ...            ...                  ...                  ...
2020-12-29 14:15:00+05:30  278.000000  278.350006  277.049988  277.399994  1333187          0             0  ...       0        1       0      70.481224      69.749927            62.894709            57.654916
2020-12-29 14:30:00+05:30  277.500000  277.899994  276.500000  277.750000  1194604          0             0  ...       1        0       0      68.110486      69.690995            54.537582            59.388621
2020-12-29 14:45:00+05:30  277.649994  277.850006  276.850006  276.950012   841302          0             0  ...       1        0       0      59.166718      65.919476            29.097088            48.843126
2020-12-29 15:00:00+05:30  277.000000  278.399994  277.000000  277.950012  2298641          0             0  ...       1        0       0      63.750203      63.675802            33.053394            38.896021
2020-12-29 15:15:00+05:30  277.950012  278.450012  277.250000  277.750000  2537262          0             0  ...       1        0       0      63.750203      62.222375            26.968237            29.706239

[75 rows x 62 columns]
root@node54310-pandas-r:~/sas#

```

**Comment by rahulmr** [comment]
Seems like a known issue with windows and multiprocessing freeze_support()
[Link1](https://stackoverflow.com/questions/13922597/multiprocessing-freeze-support)
[Link2](https://stackoverflow.com/questions/24363300/freeze-support-bug-in-using-scikit-learn-in-the-anaconda-python-distro)
[Link3](https://stackoverflow.com/questions/29690091/python2-7-exception-the-freeze-support-line-can-be-omitted-if-the-program)
[Link4](https://github.com/pytorch/pytorch/issues/7485)

**Comment by rahulmr** [comment]
You can close this issue if you do not want to investigate further, as I have got the solution which is as below:
source code
```python

import yfinance as yf
import pandas_ta as ta


def test():
    sbin = yf.Ticker("SBIN.NS")
    df = sbin.history(period="3d", interval="15m")
    df.ta.strategy("Momentum", verbose=True)
    print(df)

if __name__ == '__main__':
    test()
```
pwsh.exe console output


```
(venv) PS D:\trading\0sas_algo> cat .\crash.py
import yfinance as yf
import pandas_ta as ta


def test():
    sbin = yf.Ticker("SBIN.NS")
    df = sbin.history(period="3d", interval="15m")
    df.ta.strategy("Momentum", verbose=True)
    print(df)

if __name__ == '__main__':
    test()
(venv) PS D:\trading\0sas_algo> python .\crash.py
[+] Strategy: Momentum
[i] Indicator arguments: {'append': True}
[i] Excluded[10]: above, above_value, below, below_value, cross, cross_value, long_run, short_run, trend_return, vp
[i] Multiprocessing: 8 of 8 cores.
[i] Total indicators: 34
[i] Columns added: 55
                                 open        high         low       close   volume  Dividends  ...  STOCHd_14_3_3  STOCHRSIk_14_14_3_3  STOCHRSId_14_14_3_3  TRIX_30_9  TRIXs_30_9  TSI_13_25
datetime                                                                                       ...
2020-12-24 09:15:00+05:30  265.100006  266.549988  264.700012  266.299988  1653388          0  ...            NaN                  NaN                  NaN        NaN         NaN        NaN
2020-12-24 09:30:00+05:30  266.299988  266.649994  265.500000  266.299988  1346556          0  ...            NaN                  NaN                  NaN        NaN         NaN        NaN
2020-12-24 09:45:00+05:30  266.299988  266.450012  265.500000  266.100006   914542          0  ...            NaN                  NaN                  NaN        NaN         NaN        NaN
2020-12-24 10:00:00+05:30  266.100006  267.649994  266.100006  267.399994  1594715          0  ...            NaN                  NaN                  NaN        NaN         NaN        NaN
2020-12-24 10:15:00+05:30  267.299988  267.799988  266.649994  266.799988  1050138          0  ...            NaN                  NaN                  NaN        NaN         NaN        NaN
...                               ...         ...         ...         ...      ...        ...  ...            ...                  ...                  ...        ...         ...        ...
2020-12-29 14:15:00+05:30  278.000000  278.350006  277.049988  277.399994  1333187          0  ...      69.749927            62.894709            57.654916   1.585756    2.020604  27.573298
2020-12-29 14:30:00+05:30  277.500000  277.899994  276.500000  277.750000  1194604          0  ...      69.690995            54.537582            59.388621   1.498974    1.906078  26.838714
2020-12-29 14:45:00+05:30  277.649994  277.850006  276.850006  276.950012   841302          0  ...      65.919476            29.097088            48.843126   1.417326    1.798951  24.185578
2020-12-29 15:00:00+05:30  277.000000  278.399994  277.000000  277.950012  2298641          0  ...      63.675802            33.053394            38.896021   1.340583    1.698658  23.613085
2020-12-29 15:15:00+05:30  277.950012  278.450012  277.250000  277.750000  2537262          0  ...      62.222375            26.968237            29.706239   1.268367    1.604655  22.658487

[75 rows x 62 columns]
(venv) PS D:\trading\0sas_algo>

```

**Comment by twopirllc** [comment]
@rahulmr,

Excellent deductive work on figuring out the Issue. I have seen that wrapping it in ```__main__``` before could solve the Issue. Glad that it works though!

Regards,
KJ

**Comment by mjheydari** [comment]
Hi I have same problem with this code please help me. I use VScode and python 3.9.1 but get same error:


........ 

File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    return Popen(process_obj)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 143, in get_preparation_data
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    _check_not_importing_main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\multiprocessing\spawn.py", line 136, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
........

**Comment by nbodie13** [comment]
Thank you for these posts, and especially thank you for your workaround @rahulmr !!

**Comment by eervin123** [comment]
fwiw I am on a Mac Apple M1 Pro and experiencing the same issue. 

This workaround enables me to continue. I just wanted to post in case people thought it was a Windows only issue. 

My code was 

``` Python
import pandas as pd
import pandas_ta as ta

file_path = '~/Documents/data/hourly_BTCUSDT.csv'
df = pd.read_csv(file_path, parse_dates=True, index_col=0)

all_df = df.ta.strategy(verbose=True)
```
Note, I ran it inside a code cell in vs code using the `#%%` at the top.

I also ran the file using the "play" button at the top of the vscode editor (sorry I'm not sure what that is called). I also tested it by running it from the terminal prompt `python test.py`

The datafile is similar in structure and format as following. The only difference is that it was based on hourly data not minutely data. 

I'd get the same error regardless.  

```
Open time,Open,High,Low,Close,Volume
2022-01-01 08:00:00+00:00,47194.73,47228.86,47174.27,47174.28,7.95682
2022-01-01 08:01:00+00:00,47174.28,47174.28,47117.76,47125.88,12.63977
2022-01-01 08:02:00+00:00,47125.88,47155.13,47113.6,47131.31,24.2726
2022-01-01 08:03:00+00:00,47131.31,47154.27,47118.67,47131.0,3.81929
2022-01-01 08:04:00+00:00,47131.01,47163.19,47125.08,47153.78,12.99805
```

**Comment by twopirllc** [comment]
@eervin123,

This is a common issue from Windows users and has been addressed in the pinned Issue #181. Have you tried the solution posted there? Essentially using:

```python
if __name__ == '__main__':
    # Your code ...
```

Fwiw, I also develop on an Apple M1 Macbook so I haven't ran into this issue. But then again, I do not use code cells in VS Code for development or testing.

KJ
